File: Lab/Lab_5/Java/Lab_5_2.java
import java.util.Arrays;
import java.util.Vector;
import java.util.stream.Collectors;

public class Lab_5_2 {
    public static int MOD = 1337;

    public static int powMe(int a, int b)
    {
        int oddResult = 1;
        a %= MOD;

        while (b > 0) {
            if(b % 2 == 1)
                oddResult = (oddResult * a) % MOD;
            
            a = (a * a) % MOD;
            b /= 2;
        } 
        return oddResult;
    }
    public static int superPowMe(int a, Vector<Integer> b) {
        if(b.isEmpty())
            return 1;

        int last = b.lastElement();
        Vector<Integer> newVec = new Vector<>(b);
        newVec.remove(newVec.size() - 1);
        
        int lastPow = powMe(a, last);
        int leftPows = powMe(superPowMe(a, newVec), 10);

        return (lastPow * leftPows) % MOD;
    }

    public static int superPow(int a, int[] b) {
        Vector<Integer> myvec = new Vector<>(Arrays.stream(b).boxed().collect(Collectors.toList()));
        return superPowMe(a, myvec);
    }

    public static void main(String[] args) {

        int[] b = {1, 0};

        System.out.println(superPow(2, b));
    }
}

File: Lab/Lab_5/Cpp/Lab_5_2.cpp
#include <bits/stdc++.h>

using namespace std;

class Lab_5_2{
public:
    static const int MOD = 1337;

    int powMe(int a, int b){
        int oddResult = 1;

        a%=MOD;

        while(b > 0){
            if(b % 2 == 1)
                oddResult = (oddResult * a) % MOD;

            a = (a*a) % MOD;
            b/=2;
        }
        return oddResult;
    }

    int superPow(int a, vector<int>& b) {
        if(b.empty())
            return 1;
        
        int last = b.back();
        b.pop_back();

        int lastPow = powMe(a, last);
        int leftPow = powMe(superPow(a, b), 10);

        return (lastPow * leftPow) % MOD;
    }
};

int main()
{
    Lab_5_2 sol;

    vector <int> vec = {1, 0};

    cout << sol.superPow(2, vec);
}